chore(observability): enhance log rate limiting layer to apply the rate limits per key or per component

jszwedko opened the pull request:
Closes: #6342 

**EDIT** See addendum: https://github.com/timberio/vector/pull/6563#issuecomment-789205204

## Background

Our tracing layer automatically rate limits, via a layer in `lib/tracing-limit`, a given tracing callsite if it has an `internal_rate_limit_secs` field.

For example:

```rust
info!( message = "This message is rate limited", internal_log_rate_secs = 5);
```

Will print the following over 6 seconds:

```
  INFO basic: This message is rate limited internal_log_rate_secs=5
  INFO basic: Internal log [This message is rate limited] is being rate limited.
  INFO basic: Internal log [This message is rate limited] has been rate limited 4 times.
  INFO basic: This message is rate limited internal_log_rate_secs=5
```

Note that this is rate limiting is based solely on the callsite of the trace call so that, for example, if multiple code paths hit the same line, they will all be rate limited together. This causes problems when sharing code across components as errors in one component could mask errors in another.

## Usage

```rust
 for i in 0..40 {
            trace!("This field is not rate limited!");
            for key in &["foo", "bar"] {
                info!(
                    message = "This message is rate limited by its key",
                    count = &i,
                    internal_log_rate_secs = 5,
                    internal_log_rate_key = key,
                );
            }
            std::thread::sleep(std::time::Duration::from_millis(1000));
        }
```

Prints:

```
 TRACE by_key: This field is not rate limited!
  INFO by_key: This message is rate limited by its key count=0 internal_log_rate_secs=5 internal_log_rate_key="foo"
  INFO by_key: This message is rate limited by its key count=0 internal_log_rate_secs=5 internal_log_rate_key="bar"
 TRACE by_key: This field is not rate limited!
  INFO by_key: Internal log [This message is rate limited by its key internal_log_rate_key=foo].
  INFO by_key: Internal log [This message is rate limited by its key internal_log_rate_key=bar].
 TRACE by_key: This field is not rate limited!
 TRACE by_key: This field is not rate limited!
 TRACE by_key: This field is not rate limited!
 TRACE by_key: This field is not rate limited!
  INFO by_key: Internal log [This message is rate limited by its key] has been rate limited 4 times.
  INFO by_key: Internal log [This message is rate limited by its key] has been rate limited 4 times.
```

```rust
            trace!("This field is not rate limited!");
            for key in &["foo", "bar"] {
                let span = info_span!(
                    "sink",
                    component_kind = "sink",
                    component_name = &key,
                    component_type = "fake",
                );
                let _enter = span.enter();
                info!(
                    message = "This message is rate limited by its component",
                    count = &i,
                    internal_log_rate_secs = 5,
                );
            }
```

Prints:

```
TRACE by_component: This field is not rate limited!
  INFO sink{component_kind="sink" component_name="foo" component_type="fake"}: by_component: This message is rate limited by its component count=0 internal_log_rate_secs=5
  INFO sink{component_kind="sink" component_name="bar" component_type="fake"}: by_component: This message is rate limited by its component count=0 internal_log_rate_secs=5
 TRACE by_component: This field is not rate limited!
  INFO sink{component_kind="sink" component_name="foo" component_type="fake"}: by_component: Internal log [This message is rate limited by its component] is being rate limited.
  INFO sink{component_kind="sink" component_name="bar" component_type="fake"}: by_component: Internal log [This message is rate limited by its component] is being rate limited.
 TRACE by_component: This field is not rate limited!
 TRACE by_component: This field is not rate limited!
 TRACE by_component: This field is not rate limited!
 TRACE by_component: This field is not rate limited!
  INFO sink{component_kind="sink" component_name="foo" component_type="fake"}: by_component: Internal log [This message is rate limited by its component] has been rate limited 4 times.
  INFO sink{component_kind="sink" component_name="bar" component_type="fake"}: by_component: Internal log [This message is rate limited by its component] has been rate limited 4 times.
```

## Implementation

This PR introduces an enhancement to the tracing rate limiting layer that we already have to be able to rate limit events separately, for the same callsite, based either on specific event key value, `internal_rate_limit_key`, or automatically if there is a `component_name` associated a span that the event is within.

I believe the automatic partitioning of rate limits by component name unblock https://github.com/timberio/vector/pull/6357 which moves all components using templates to using a single internal event and, thus, a single trace callsite.

I believe the ability to partition rate limits by a key specified on the event will unblock #6307 which aims to allow rate limiting the VRL `log` function.

I think it is probably easiest to review each commit in order. I might also start with looking at the refactored + added examples in the `lib/tracing-limit/examples` directory.

* fb3b5c4 simply refactors the current tracing layer to, instead of wrapping a fmt subscriber, which is the bit that outputs the log messages wrap a fmt layer. This is required because adding the ability to rate limit by the, to be added in a following commit, `internal_rate_limit_key`, requires accessing the event data which is only available on `on_event` which only receives a reference to the event. Thus the new implementation introduces a sort of rate limiting middleware which receives the `on_event` call first and conditionally forwards to the next layer
* 0588f03 finishes this refactoring by actually moving the logic to determine when the next layer receives the event into `on_event` rather than returning true/false conditionally from `enabled`
* 115e940 adds the ability to specify `internal_rate_limit_key` to have log events separately rate limited based on the value of this key
* a1b2787 just drops the callsite store from the rate limiting layer as the metadata is already available on the event so we don't need to fetch it from the callsite
* d6be92b introduces automatic rate limiting if `component_name` appears in a span attribute. This will automatically separately rate limit logs based on their component
* 8739807 clippy

Alternatives:

* We could drop the `internal_log_rate_key` support and simply rely on trace span values. I added `internal_log_rate_key` because I thought that the `remap` use-case for the `log` function would need it, but it seems like it could also just add a span attribute that is the VRL script line number and we could add support to the tracing layer to differentiate on this as well. This could maybe be done dynamically by letting developers specify, on the event, which span attributes should be considered for separately rate limiting. Something like: `info!( message = "This message is rate limited", internal_log_rate_secs = 5, internal_log_rate_limit_by = ["component_name", "remap_line_number"]);`.

TODO:
- [x] Tests? The didn't exist before, but seem like a good idea.
- [x] Update and run benches

jszwedko commented on lib/tracing-limit/src/lib.rs:
```diff
@@ -10,165 +10,200 @@ use std::{
-                Some(&rate_limit as &dyn Value),
-            )];
+    // TODO is this right?
+    _subscriber: std::marker::PhantomData<S>,
```
I'm not familiar at all with this `PhantomData` marker. I simply followed rustc's recommendation to add it.

bruceg replied:
All generic parameters have to be used in the struct. Since there is no actual data to correspond with `S`, you can add some `PhantomData` with that type. It's a zero-sized structure element that satisfies that use requirement.

blt replied:
You'll also see these in structs where the data is there but Rust can't prove that it is. For instance, if you're doing `unsafe` style programming and you have a raw pointer to a something in memory that your struct "owns" but was not created by Rust (or was "forgotten" with with `mem::forget`). If your struct is parameterized on the type of that thing you need to use `PhantomData` to square the compiler's internal logic. 

Fun stuff.

jszwedko commented on src/trace.rs:
```diff
@@ -44,30 +44,32 @@ pub fn init(color: bool, json: bool, levels: &str) {
+    let subscriber = tracing_subscriber::registry::Registry::default()
+        .with(tracing_subscriber::filter::EnvFilter::from(levels));
+
     let dispatch = if json {
```
I wanted to try to refactor this because there is a quite a bit of duplication, but I struggled to satisfy the complier due to all of the generics used in tracing-subscriber. This duplication is already there in the current implementation though.

FungusHumungus replied:
I'd be very interested in eavesdropping on that if the timezones can work..

jszwedko replied:
👍 I would definitely learn from pairing on it.

jszwedko replied:
Just dropping a note for posterity that we attempted to refactor this, but it quickly grew complex enough that it didn't seem worth it for the small amount of duplication (we started adding in a new wrapping, generic, type around the fmt layers).

jszwedko commented on src/trace.rs:
```diff
@@ -110,29 +112,29 @@ pub fn subscribe() -> TraceSubscription {
-struct BroadcastSubscriber<F> {
-    formatter: F,
+struct BroadcastSubscriber<S> {
+    subscriber: S,
```
Just renamed this to be more clear.

blt reviewed (approved):
Absent any CI dings or what not this seems good to my eye. I will note that you have a "TODO" comment still yet, but I imagine you're aware of that.

juchiast commented on lib/tracing-limit/src/lib.rs:
```diff
@@ -178,39 +213,125 @@ where
 
+#[derive(Debug)]
+struct State {
+    start: Option<Instant>,
```
This `Option` isn't necessary anymore because it's always `Some` now.

jszwedko replied:
Indeed, thank you 😄

juchiast reviewed (commented):
Thanks! This should unblock #6242 .

juchiast commented on lib/tracing-limit/src/lib.rs:
```diff
@@ -10,165 +10,200 @@ use std::{
+                    self.create_event(&ctx, metadata, message, state.limit);
+                }
+            } else {
+                // done rate limiting
```
I think we have to call `inner.on_event` here, for example, if `log_rate_secs` is 1 and another event emitted 2 seconds later, it would be rate-limited.

jszwedko replied:
Good catch, I'll look into that. I also realized there could be a race condition here too if two log events come in at "same time" where they both check for state, find none, output, and insert into the state map.

juchiast commented on lib/tracing-limit/src/lib.rs:
```diff
@@ -10,165 +10,200 @@ use std::{
-                limit: DEFAULT_LIMIT,
-                message: String::new(),
-            };
+            Interest::sometimes()
```
This is a minor issue and I'm not sure if I'm correct, `Interest::sometimes` means: `enabled` will be called on each event, if it returns true then `on_event` will be called, otherwise don't call it. 

Because this layer just calls `inner.enabled` and doesn't do anything else, it should also return `inner.register_callsite` here.

jszwedko replied:
Good catch, you are right, this should just proxy to inner.

FungusHumungus commented on lib/tracing-limit/src/lib.rs:
```diff
@@ -10,165 +10,200 @@ use std::{
+
+        let id = RateKeyIdentifier {
+            callsite: metadata.callsite(),
+            rate_limit_key: limit_visitor.key.clone(),
```
I suspect this code is on a hot path. I wonder if it would be beneficial to avoid these clones here by making `RateKeyIdentifier` something like:

```rust
struct RateKeyIdentifier<S: AsRef<str>> {
    callsite: Identifier,
    rate_limit_key: Option<S>,
    component_name: Option<S>,
}
```

Then we would only need to clone the strings when storing the Identifiers.

jszwedko replied:
👍 I was thinking about that too. I'll give it a shot.

jszwedko replied:
This clone didn't appear to make much of an impact on the benches (see https://github.com/timberio/vector/pull/6563#issuecomment-789205204) so I just left it. I welcome thoughts though!

FungusHumungus commented on lib/tracing-limit/src/lib.rs:
```diff
@@ -178,39 +212,125 @@ where
         }
     }
 
     fn record_str(&mut self, field: &Field, value: &str) {
```
It seems `record_str` is only invoked for fields that are `&str`.  If you changed the `by_key.rs` example to be:

```rust
            for key in &[String::from("foo"), String::from("bar")] {
             ....
                    internal_log_rate_key = %key,
```

then they won't be used as a rate limit key. If we also override `record_debug` and duplicate this functionality, then it would work.


Without `record_debug` you would still be able to get it working by coercing the string to a str like:

```rust
            for key in &[String::from("foo"), String::from("bar")] {
             ....
                    internal_log_rate_key = key.as_str(),
```

But, I bet people would always forget this, and there is no way for the code to alert that this has happened...

jszwedko replied:
It looks like you can't actually pass a `String` there, it has to be a `tracing::Value`: https://docs.rs/tracing/0.1.10/tracing/trait.Value.html

jszwedko replied:
This does limit the usage to string values though (as opposed to, for example, an integer. I think I may actually back out this particular change though as I think the implementation that rate limits just on span context fields should be sufficient for both of our current needs: rate limiting in remap and sharing internal events.

FungusHumungus replied:
> It looks like you can't actually pass a `String` there,

You can pass a String - or any other type, but in the macro you prepend it with `%`, which essentially takes the `Display` value. Prepending with `?` takes the `Debug` value. 

It seems that there is no `record_display` method, not sure why, - both types go through the `record_debug` call. The default implementation for the other types just calls `record_debug`.

Not sure I understand what you mean here?

> This does limit the usage to string values

jszwedko replied:
Aha, I missed your introduction of the `%`. Yeah, I see.

For

> This does limit the usage to string values

I just mean that the visitor that records the value for the key is currently only looking at `&str`s via `visit_str` so if, for example, a `uint` was passed, it wouldn't pick that up.

I still think I'm going to back out this specific change of allowing a key, in-lieu of using tracing span data, but I'll carry over your suggestion to the tracing span visitor.

I did want to check with you to see if just using span data would be sufficient for the remap case though. I was thinking that remap could add a span that records the source line number. In addition to being able to use this to group the log messages so they are rate limited separately, I also thought this would be a useful enhancement to have internal vector logs also be tagged with the remap source line number to aid with debugging. Maybe we could pair on this this week?

jszwedko replied:
I implemented `record_debug` for the `message` field. I also discussed with Stephen offline that the span approach should be sufficient for VRL too so I dropped the implementation of `internal_rate_limit_key`.

FungusHumungus reviewed (changes requested):
I think we need to also `record_debug`..

jszwedko commented:
Additional changes, which can probably be viewed at once (via https://github.com/timberio/vector/pull/6563/files/873980733866c297a7d371baff1b0fe581b2e029..3e6b389098e83da39292a1a122febba6f56e2c04):

* Replaced `RwLock<HashMap<_,_>>` with `DashMap<_,_>` after a brief interlude into just using a `Mutex`. This was part of some refactoring I did with @blt which I think ended up making the code much more legible. This also allowed me to drop the use of atomics as we maintain a log on the entry.
* Removed the support for `internal_log_rate_limit_key` grouping as it doesn't seem like we'll need this; instead we can just rely on the span data to differentiate and rate limit the same callsite separately.
* Fixed the benches and added some tests.

I ran some benches that show this being roughly as fast as `master` in the pathological case of every event flowing through vector resulting in a rate limited log message:

Master:

```
➜  vector git:(master) perf stat -r 10 bash -c 'yes "invalid json" | head -n 1000000 | ./target/release/vector-081372e15 --config /tmp/perf.toml > /dev/null'

 Performance counter stats for 'bash -c yes "invalid json" | head -n 1000000 | ./target/release/vector-081372e15 --config /tmp/perf.toml > /dev/null' (10 runs):

     165016.838937      task-clock:u (msec)       #    7.116 CPUs utilized            ( +-  0.54% )
                 0      context-switches:u        #    0.000 K/sec
                 0      cpu-migrations:u          #    0.000 K/sec
              6652      page-faults:u             #    0.040 K/sec                    ( +-  1.73% )
   <not supported>      cycles:u
   <not supported>      instructions:u
   <not supported>      branches:u
   <not supported>      branch-misses:u

      23.190910924 seconds time elapsed                                          ( +-  0.69% )
```

Recent commit here:

```
perf stat -r 10 bash -c 'yes "invalid json" | head -n 1000000 | ./target/release/vector-f224c6441 --config /tmp/perf.toml > /dev/null'

 Performance counter stats for 'bash -c yes "invalid json" | head -n 1000000 | ./target/release/vector-f224c6441 --config /tmp/perf.toml > /dev/null' (10 runs):

     169876.031647      task-clock:u (msec)       #    7.202 CPUs utilized            ( +-  0.56% )
                 0      context-switches:u        #    0.000 K/sec
                 0      cpu-migrations:u          #    0.000 K/sec
              6361      page-faults:u             #    0.037 K/sec                    ( +-  0.79% )
   <not supported>      cycles:u
   <not supported>      instructions:u
   <not supported>      branches:u
   <not supported>      branch-misses:u

      23.587183886 seconds time elapsed                                          ( +-  0.62% )
```

jszwedko commented:
I think this is ready for review again @blt @FungusHumungus @juchiast

blt commented on lib/tracing-limit/src/lib.rs:
```diff
@@ -1,210 +1,297 @@
             }
+            self.inner.on_event(event, ctx);
+            state.start = Instant::now();
+            state.count = 1;
```
I think this should be set to 0? If we're resetting the interior state back to default then we need to be zero, unless I'm missing something. Else the next time we come back around here the count will be set to 2, on account of line 150?

jszwedko replied:
I also found this a bit confusing :)

We want to set it to 1 here because we just emitted the event above so the next one that comes in should be rate limited (thus we want it to be 2 the next time around).

blt replied:
Aaaaah, yes. Makes perfect sense. I think if you found it confusing and I found it confusing I'd appreciate a comment explaining this.

juchiast reviewed (approved):
Clippy failed, otherwise look good!

jszwedko commented:
@FungusHumungus I'll merge this given the two approvals, but I'll make the enhancements we discussed to rate limit on the additional span attributes in a follow-up PR to make it easier to review in isolation.
